grpc_unary_attestation/src/server.rs

```diff
@@ -18,20 +18,88 @@
 //! protocol.
 
 use crate::proto::{unary_session_server::UnarySession, UnaryRequest, UnaryResponse};
+use lru::LruCache;
+use oak_remote_attestation::handshaker::{AttestationBehavior, Encryptor, ServerHandshaker};
 use oak_utils::LogError;
+use std::sync::{Arc, Mutex};
 use tonic;
 
-/// gRPC Attestation Service implementation.
-pub struct AttestationServer<F, L: LogError> {
+type SessionId = u64;
+enum SessionState {
+    // Boxed due to large size difference, ref: https://rust-lang.github.io/rust-clippy/master/index.html#large_enum_variant
+    HandshakeInProgress(Box<ServerHandshaker>),
+    EncryptedMessageExchange(Encryptor),
+}
+
+/// Maintains remote attestation state for a number of sessions
+struct SessionsTracker {
     /// PEM encoded X.509 certificate that signs TEE firmware key.
-    _tee_certificate: Vec<u8>,
-    /// Processes data from client requests and creates responses.
-    _request_handler: F,
+    tee_certificate: Vec<u8>,
     /// Configuration information to provide to the client for the attestation step.
-    _additional_info: Vec<u8>,
+    additional_info: Vec<u8>,
+    known_sessions: LruCache<SessionId, SessionState>,
```

Comment:
Note that this is pretty much already what is happening, with the exception that the Mutex is held while normalizing SessionState (which is v cheap). The mutex isn't locked while processing a request, but locked separately for retrieval & return.

Reply:
I know it is quite cheap, but wouldn't agree that it is very cheap. Cloning the additional info, which could be large, and handling errors by generating strings from the error details (which could also be large) might take some time. It would be preferable not to have such an important mutex locked for that entire time.